hancock/routes.py

```diff
@@ -61,37 +66,58 @@ class FetchUrl(Resource):
     @api.marshal_with(url_resource)
     @jwt_required()
     def post(self):
-      S3Operations.client_options()
-      response = S3Operations.generate_presigned_url(Bucket=api.payload['Bucket'], Key=api.payload['Key'])
-
-      return response
+        S3Operations.client_options()
+        response = S3Operations.generate_presigned_url(Bucket=api.payload['Bucket'], Key=api.payload['Key'],
+                                                       Expiration=EXPIRATION)
+        return response
 
 @api.route('/receive_async_messages')
+
 class ReceiveAsyncMessages(Resource):
     @jwt_required()
     @api.expect(message_resource)
-    @api.marshal_with(url_resource, as_list=True)
+    @api.response(200, 'Job Complete')
     def post(self):
-        print(f"message received:{api.payload['async_message']}")
-        payload = ast.literal_eval(api.payload['async_message'])
-        datasetList = payload["datasetList"]
-        output_ls = []
-        for item in datasetList:
-             output = get_associated_payload(item['pid'])
-             output_ls.append(output)
-        url_ls=[]
-        if output_ls:
-            for output in output_ls:
-                bucket = urlparse(output[0]['sourceFolderHost'])[1].split('.')[0]
-                key = output[0]['sourceFolder'].strip('/')
-                S3Operations.client_options()
-                url = S3Operations.generate_presigned_url(Bucket=bucket, Key=key)
-                url_ls.append(url[0])
+        app.logger.info(f"message received:{api.payload['async_message']}")
+        try:
+            payload = ast.literal_eval(api.payload['async_message'])
+            dataset_list = payload["datasetList"]
+        except (SyntaxError, ValueError) as e:
+            app.logger.debug(e)
+            abort(401, "Invalid Message")
+
+        if dataset_list:
+            output_ls = [get_associated_payload(item['pid']) for item in dataset_list]
+
+            url_ls = []
+            if output_ls:
+                for output in output_ls:
+                    try:
+                        bucket, key = check_process_bucket_key(output[0])
+                        S3Operations.client_options()
+                        url = S3Operations.generate_presigned_url(Bucket=bucket, Key=key)
+                        url_ls.append(url[0])
+                    except Exception as e:
+                        app.logger.debug(e)
+                        continue
+                    finally:
```

Comment:
I am not sure you want to use `finally` here as this will get executed even when an exception is caught, just before `continue` is called. Perhaps `else` is more appropriate?

Reply:
yes you are right. I think this is a hangover from an earlier implementation. Will change